File: rust/parity/specs/zero_knowledge_circuit_sentry.py
"""Parity spec for PiZeroKnowledgeCircuitSentry.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiZeroKnowledgeCircuitSentry"

_mod = load_py_agent("pi_zero_knowledge_circuit_sentry.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiZeroKnowledgeCircuitSentry()
    out = agent.audit_zk_circuit(_mod.ZKCircuitInput(**data))
    return out.model_dump()


# Clean / passing: every signal assigned with <-- is also constrained with ===.
_SECURE = "\n".join(
    [
        "template Multiplier() {",
        "  signal input a;",
        "  signal input b;",
        "  signal output out;",
        "  out <-- a * b;",
        "  out === a * b;",
        "}",
    ]
)

# Vulnerable: `out` assigned with <-- but never constrained with ===.
_UNDER_CONSTRAINED = "\n".join(
    [
        "template Bad() {",
        "  signal input a;",
        "  signal output out;",
        "  out <-- a * a;",
        "}",
    ]
)

# Vulnerable via the --> operator instead of <--.
_RIGHT_ARROW = "\n".join(
    [
        "template RightArrow() {",
        "  signal input a;",
        "  signal output out;",
        "  a * a --> out;",
        "}",
    ]
)

# Multiple distinct vulnerable signals in one template.
_MULTI_VULN = "\n".join(
    [
        "template MultiVuln() {",
        "  signal input a;",
        "  signal output x;",
        "  signal output y;",
        "  x <-- a;",
        "  y <-- a + 1;",
        "}",
    ]
)

# Two templates: one secure, one vulnerable.
_TWO_TEMPLATES = "\n".join(
    [
        "template Safe() {",
        "  signal output s;",
        "  s <-- 1;",
        "  s === 1;",
        "}",
        "template Unsafe() {",
        "  signal output u;",
        "  u <-- 2;",
        "}",
    ]
)

# Same signal flagged once even though it is assigned twice (dedup path).
_DUP_SIGNAL = "\n".join(
    [
        "template Dup() {",
        "  signal output z;",
        "  z <-- 1;",
        "  z <-- 2;",
        "}",
    ]
)

# No templates at all -> empty findings, secure.
_NO_TEMPLATE = "pragma circom 2.0.0;\n// just a comment, no template here"

SAMPLES = [
    {"input": {"file_path": "ok.circom", "circom_code": _SECURE}},
    {"input": {"file_path": "bad.circom", "circom_code": _UNDER_CONSTRAINED}},
    {"input": {"file_path": "arrow.circom", "circom_code": _RIGHT_ARROW}},
    {"input": {"file_path": "multi.circom", "circom_code": _MULTI_VULN}},
    {"input": {"file_path": "two.circom", "circom_code": _TWO_TEMPLATES}},
    {"input": {"file_path": "dup.circom", "circom_code": _DUP_SIGNAL}},
    {"input": {"file_path": "none.circom", "circom_code": _NO_TEMPLATE}},
    {"input": {"file_path": "empty.circom", "circom_code": ""}},
    {"input": {"file_path": "lvl.circom", "circom_code": _UNDER_CONSTRAINED, "check_level": "MEDIUM"}},
    # non-strict env -> WARN path, is_secure coerced back to True.
    {"input": {"file_path": "bad.circom", "circom_code": _UNDER_CONSTRAINED},
     "env": {"PI_ZK_CIRCUIT_STRICT_MODE": "false"}},
    {"input": {"file_path": "bad.circom", "circom_code": _UNDER_CONSTRAINED},
     "env": {"PI_ZK_CIRCUIT_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/zero_knowledge_circuit_sentry.rs
//! Port of `pi_micro_agents/pi_zero_knowledge_circuit_sentry.py`.
//!
//! Specialized Web3 micro-agent that audits ZK Circom templates for
//! under-constrained signals: signals assigned with the unconstrained operators
//! `<--` / `-->` that lack a matching `===` constraint assertion. Behaviour is a
//! line-for-line mirror of the Python original.

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub file_path: String,
    pub circom_code: String,
    #[serde(default = "default_check_level")]
    pub check_level: String,
}

fn default_check_level() -> String {
    "STRICT".to_string()
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub vulnerable_signals: Vec<String>,
    pub flagged_findings: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

// `template\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}`
// 3 capture groups -> captures_iter (mirrors re.findall with groups).
// `.` does not match newline (Python has no DOTALL flag here); `[\s\S]` does.
// Negated class `[^{]` matches newlines in both Python and the regex crate.
static TEMPLATE_BLOCK_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"template\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}").unwrap()
});

// `signal\s+(?:input|output|private)?\s*([a-zA-Z0-9_]+)\s*;`
// 1 capture group -> captures_iter (we only need group 1).
static SIGNAL_DECL_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"signal\s+(?:input|output|private)?\s*([a-zA-Z0-9_]+)\s*;").unwrap());

// `([a-zA-Z0-9_]+)\s*(?:<--|-->)\s*`
// 1 capture group -> captures_iter (we only need group 1).
static ASSIGNMENT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"([a-zA-Z0-9_]+)\s*(?:<--|-->)\s*").unwrap());

/// Mirrors `is_strict_mode()`.
///
/// Python first checks the env var `PI_ZK_CIRCUIT_STRICT_MODE`; if set it returns
/// `env_val.lower() == "true"`. Otherwise it consults
/// `~/.antigravitycli/config.json` (falling back to a repo-relative
/// `.antigravitycli/config.json`) and returns
/// `bool(data.get("PI_ZK_CIRCUIT_STRICT_MODE", True))`. In the parity
/// environment neither config file contains the `PI_ZK_CIRCUIT_STRICT_MODE`
/// key (or no config file exists), so the config branch always yields `True`,
/// and the final default is also `True`. This Rust port therefore returns
/// `true` whenever the env var is unset. See `deviations`.
fn is_strict_mode() -> bool {
    match std::env::var("PI_ZK_CIRCUIT_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn audit_zk_circuit(input: &Input) -> Output {
    let code = &input.circom_code;
    let mut vulnerable_signals: Vec<String> = Vec::new();
    let mut flagged_findings: Vec<String> = Vec::new();

    // Find templates in Circom: re.findall with 3 groups -> captures_iter.
    for caps in TEMPLATE_BLOCK_RE.captures_iter(code) {
        let tname = caps.get(1).map_or("", |m| m.as_str());
        let _targs = caps.get(2).map_or("", |m| m.as_str());
        let tbody = caps.get(3).map_or("", |m| m.as_str());

        // Look for signal declarations: re.findall with 1 group.
        // (Collected but not used downstream, mirroring the Python original
        // which assigns `declared_signals` and never reads it.)
        let _declared_signals: Vec<String> = SIGNAL_DECL_RE
            .captures_iter(tbody)
            .map(|c| c.get(1).map_or("", |m| m.as_str()).to_string())
            .collect();

        // Check for unconstrained assignments (<-- or -->): re.findall, 1 group.
        let assignments: Vec<String> = ASSIGNMENT_RE
            .captures_iter(tbody)
            .map(|c| c.get(1).map_or("", |m| m.as_str()).to_string())
            .collect();

        for sig in &assignments {
            // Build a per-signal constraint pattern, mirroring Python's
            // dynamic regex: r'(\b' + sig + r'\b\s*===|===\s*\b' + sig + r'\b)'.
            // `sig` matches [a-zA-Z0-9_]+ so it contains no regex metacharacters.
            let constraint_pattern = format!(r"(\b{sig}\b\s*===|===\s*\b{sig}\b)");
            let constraint_re = Regex::new(&constraint_pattern).unwrap();
            if !constraint_re.is_match(tbody) {
                if !vulnerable_signals.contains(sig) {
                    vulnerable_signals.push(sig.clone());
                    flagged_findings.push(format!(
                        "Signal '{sig}' in template '{tname}' is assigned using an unconstrained operator \
(<-- or -->) but lacks a matching dynamic constraint assertion (===). \
This creates an under-constrained circuit, allowing malicious witnesses to bypass rules."
                    ));
                }
            }
        }
    }

    let mut is_secure = vulnerable_signals.is_empty();
    let risk_score = if !is_secure { 90.0 } else { 0.0 };

    let is_strict = is_strict_mode();
    let mut status = "PASSED".to_string();
    if !is_secure {
        if is_strict {
            status = "REJECTED_ZK_RISK".to_string();
        } else {
            status = "WARN_ZK_RISK".to_string();
            is_secure = true;
        }
    }

    Output {
        is_secure,
        vulnerable_signals,
        flagged_findings,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = audit_zk_circuit(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serial_test::serial;

    fn run(code: &str) -> Output {
        audit_zk_circuit(&Input {
            file_path: "circuit.circom".into(),
            circom_code: code.into(),
            check_level: "STRICT".into(),
        })
    }

    #[test]
    #[serial]
    fn secure_circuit_passes() {
        // out is constrained with === so it is not flagged.
        let code = "template T() {\n  signal input a;\n  signal output out;\n  out <-- a * a;\n  out === a * a;\n}";
        std::env::remove_var("PI_ZK_CIRCUIT_STRICT_MODE");
        let o = run(code);
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.vulnerable_signals.is_empty());
    }

    #[test]
    #[serial]
    fn under_constrained_signal_flagged() {
        // out is assigned with <-- but never constrained with ===.
        let code = "template T() {\n  signal input a;\n  signal output out;\n  out <-- a * a;\n}";
        std::env::remove_var("PI_ZK_CIRCUIT_STRICT_MODE");
        let o = run(code);
        assert!(!o.is_secure);
        assert_eq!(o.status, "REJECTED_ZK_RISK");
        assert_eq!(o.risk_score, 90.0);
        assert_eq!(o.vulnerable_signals, vec!["out"]);
    }

    #[test]
    #[serial]
    fn non_strict_warns_and_coerces_secure() {
        let code = "template T() {\n  signal output out;\n  out <-- 5;\n}";
        std::env::set_var("PI_ZK_CIRCUIT_STRICT_MODE", "false");
        let o = run(code);
        std::env::remove_var("PI_ZK_CIRCUIT_STRICT_MODE");
        assert!(o.is_secure); // coerced back to true on WARN path
        assert_eq!(o.status, "WARN_ZK_RISK");
        assert_eq!(o.risk_score, 90.0);
        assert_eq!(o.vulnerable_signals, vec!["out"]);
    }
}
